packages/hardhat-test-helpers/src/utils.ts

```diff
@@ -0,0 +1,88 @@
+import type { EIP1193Provider } from "hardhat/types";
+
+import type { NumberLike } from "./types";
+
+async function checkIfHardhatNetwork(
+  provider: EIP1193Provider
+): Promise<boolean> {
+  try {
+    // we call a method that only exists in the hardhat network
+    await provider.request({
+      method: "hardhat_getAutomine",
+    });
+
+    return true;
+  } catch (e) {
+    return false;
+  }
+}
+
+export async function getHardhatProvider(): Promise<EIP1193Provider> {
+  try {
+    const hre = await import("hardhat");
+
+    const provider = hre.network.provider;
+
+    const isHardhatNetwork = await checkIfHardhatNetwork(provider);
+
+    if (!isHardhatNetwork) {
+      throw new Error(
+        "[hardhat-test-helpers] This helper can only be used on the hardhat network"
+      );
+    }
+
+    return hre.network.provider;
+  } catch (e: any) {
+    if (e.code === "MODULE_NOT_FOUND") {
+      throw new Error(
+        "[hardhat-test-helpers] These helpers can only be used inside a hardhat project"
+      );
+    }
+
+    throw e;
+  }
+}
+
+export function toRpcQuantity(x: NumberLike): string {
+  let hex: string;
+  if (typeof x === "number" || typeof x === "bigint") {
+    // TODO: check that number is safe
+    hex = `0x${x.toString(16)}`;
+  } else if (typeof x === "string") {
+    if (!x.startsWith("0x")) {
+      throw new Error(
+        "[hardhat-test-helpers] Only 0x-prefixed hex-encoded strings are accepted"
+      );
+    }
+    hex = x;
+  } else if ("toHexString" in x) {
+    hex = x.toHexString();
+  } else if ("toString" in x) {
+    hex = x.toString(16);
+  } else {
+    throw new Error(
+      `[hardhat-test-helpers] ${x} cannot be converted to an RPC quantity`
+    );
+  }
+
+  return hex.startsWith("0x") ? hex.replace("0x0", "0x") : `0x${hex}`;
```

Comment:
Something I forgot to mention: this doesn't work with `0` because it results in sending `0x` instead of `0x0`, which is invalid. Totally my fault of course, I was the one that wrote this method.

Reply:
What's the reason for this line at all? I'm trying to understand what the purpose is so I can rewrite the logic, but I'm not sure what exactly it's trying to achieve.

Reply:
The context is [this issue](https://github.com/NomicFoundation/hardhat/issues/1585).

When you use ethers's `.toHexString`, the resulting string can be left-padded with an extra zero to make the number of characters even. This means that some values can be something like `0x01` instead of `0x1`. The spec of the JSON-RPC layer [doesn't accept this](https://github.com/ethereum/execution-apis/blob/cd7f22ee84be0b18eab91cf0f00f40af75bd55b1/src/schemas/base-types.json#L42-L46), so Hardhat throws an error to be compliant.